File: src/tree_parity_machine/TreeParityMachine.java
package tree_parity_machine;

import learning_algorithm.LearningParadigm;
import learning_algorithm.Training;
import tree_parity_machine.layer.HiddenLayer;
import tree_parity_machine.layer.OutputLayer;
import tree_parity_machine.neuron.Neuron;

import java.util.Arrays;

public class TreeParityMachine implements Training {

    private int n;
    private int k;
    private int leftBound;
    private int rightBound;
    private HiddenLayer hiddenLayer;
    private OutputLayer outputLayer;
    private LearningParadigm paradigm;

    public TreeParityMachine(int n, int k, int leftBound, int rightBound, LearningParadigm learningParadigm) {
        this.n = n;
        this.k = k;
        this.leftBound = leftBound;
        this.rightBound = rightBound;
        this.paradigm = learningParadigm;
        hiddenLayer = new HiddenLayer(n, k, leftBound, rightBound, paradigm);
        outputLayer = new OutputLayer(k);
    }

    public short getOutput(short[] input) {
        return outputLayer.getOutput(hiddenLayer.getOutput(input));
    }

    @Override
    public void train(short[] input, short output) {
        Neuron[] hiddenNeurons = hiddenLayer.getNeurons();
        for (int i = 0; i < k; i++)
            hiddenNeurons[i].changeWeights(Arrays.copyOfRange(input, n * i, n * (i + 1)), output);
    }

    public short[] getSecretKey() {
        short[] key = new short[n * k];
        Neuron[] neurons = hiddenLayer.getNeurons();
        for (int i = 0; i < k; i++) {
            short[] mas = neurons[i].getWeights();
            if (n >= 0) System.arraycopy(mas, 0, key, i * n, n);
        }
        return key;
    }

    public LearningParadigm getLearningParadigm() {
        return paradigm;
    }

    public void setLearningParadigm(LearningParadigm paradigm) {
        this.paradigm = paradigm;
    }

    public int[] getTPMParams() {
        return new int[]{n, k};
    }

    @Override
    public String toString() {
        return "TreeParityMachine{" +
                "n=" + n +
                ", k=" + k +
                ", leftBound=" + leftBound +
                ", rightBound=" + rightBound +
                ", hiddenLayer=" + hiddenLayer +
                ", outputLayer=" + outputLayer +
                ", paradigm=" + paradigm +
                '}';
    }
}

File: src/main/kotlin/com/jvvladimir/machine/tree_machine/TreeParityMachine.kt
package com.jvvladimir.machine.tree_machine

import com.jvvladimir.machine.learning.LearningParadigm
import com.jvvladimir.machine.learning.Training
import com.jvvladimir.machine.tree_machine.layer.HiddenLayer
import com.jvvladimir.machine.tree_machine.layer.OutputLayer

class TreeParityMachine(
        block: TreeParityMachine.(TreeParityMachine) -> Unit
) : Training {

    init {
        this.block(this)
    }

    var n: Int = 3
    var k: Int = 4
    var leftBound: Int = -2
    var rightBound: Int = 2
    var learningParadigm: LearningParadigm = LearningParadigm.HEBBIAN

    private val hiddenLayer = HiddenLayer(n, k, leftBound, rightBound, learningParadigm)
    private val outputLayer = OutputLayer(k)

    fun getOutput(input: IntArray): Int {
        return outputLayer.getOutput(hiddenLayer.getOutput(input))
    }

    override fun train(input: IntArray, output: Int) {
        val hiddenNeurons = hiddenLayer.neurons
        for (i in 0 until k) hiddenNeurons[i].changeWeights(input.copyOfRange(n * i, n * (i + 1)), output)
    }

    val secretKey: IntArray
        get() {
            val key = IntArray(n * k)
            val neurons = hiddenLayer.neurons
            for (i in 0 until k) {
                val mas = neurons[i].weights
                if (n >= 0) System.arraycopy(mas, 0, key, i * n, n)
            }
            return key
        }

    val tpmParams: IntArray
        get() = intArrayOf(n, k)

    override fun toString() =
            "TreeParityMachine{ n= $n k= $k leftBound= $leftBound rightBound= $rightBound hiddenLayer= $hiddenLayer outputLayer= $outputLayer paradigm= $learningParadigm}"
}

fun tpm(block: TreeParityMachine.(TreeParityMachine) -> Unit): TreeParityMachine {
    return TreeParityMachine(block)
}